src/cluster/etcd/watchmanager/manager.go

```diff
@@ -71,27 +71,39 @@ func (w *manager) watchChanWithTimeout(key string, rev int64) (clientv3.WatchCha
 
 	ctx, cancelFn := context.WithCancel(clientv3.WithRequireLeader(context.Background()))
 
-	var watchChan clientv3.WatchChan
+	var (
+		watcher   = clientv3.NewWatcher(w.opts.Client())
+		watchChan clientv3.WatchChan
+	)
 	go func() {
 		wOpts := w.opts.WatchOptions()
 		if rev > 0 {
 			wOpts = append(wOpts, clientv3.WithRev(rev))
 		}
-		watchChan = w.opts.Watcher().Watch(
+		watchChan = watcher.Watch(
 			ctx,
 			key,
 			wOpts...,
 		)
 		close(doneCh)
 	}()
 
-	timeout := w.opts.WatchChanInitTimeout()
+	var (
+		timeout       = w.opts.WatchChanInitTimeout()
+		cancelWatchFn = func() {
+			cancelFn()
+			if err := watcher.Close(); err != nil {
+				w.logger.Info("error closing watcher", zap.Error(err))
+			}
+		}
+	)
+
 	select {
 	case <-doneCh:
-		return watchChan, cancelFn, nil
+		return watchChan, cancelWatchFn, nil
 	case <-time.After(timeout):
-		cancelFn()
-		return nil, nil, fmt.Errorf("etcd watch create timed out after %s for key: %s", timeout.String(), key)
+		err := fmt.Errorf("etcd watch create timed out after %s for key: %s", timeout.String(), key)
```

Comment:
can you call cancelWatchFn() here instead of having user to call this on err != nil?